### PR TITLE
Move event processing to separate threads and add event processing timer in RDS source

### PR DESCRIPTION
**Pull request by oeyh**
### Description
Move event processing to separate threads and add event processing timer in RDS source. This will help with performance testing.
 
### Issues Resolved
N/A
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java [hunk 1]
```diff
@@ -59,17 +63,20 @@ public class BinlogEventListener implements BinaryLogClient.EventListener {
...
     private final BinaryLogClient binaryLogClient;
-    private final BufferAccumulator<Record<Event>> bufferAccumulator;
+    private final Buffer<Record<Event>> buffer;
     private final List<String> tableNames;
     private final String s3Prefix;
     private final boolean isAcknowledgmentsEnabled;
     private final PluginMetrics pluginMetrics;
     private final List<Event> pipelineEvents;
     private final StreamCheckpointManager streamCheckpointManager;
+    private final ExecutorService executorService;
```
Comment (dlvenable) [inline data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java:73]: It may be good to give this a name indicating what we use it for. e.g. `binlogEventExecutorService`.

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java [hunk 2]
```diff
@@ -223,6 +233,8 @@ private void handleRowChangeEvent(com.github.shyiko.mysql.binlog.event.Event eve
         final List<String> primaryKeys = tableMetadata.getPrimaryKeys();
         final long eventTimestampMillis = event.getHeader().getTimestamp();
 
+        final BufferAccumulator<Record<Event>> bufferAccumulator = BufferAccumulator.create(buffer, DEFAULT_BUFFER_BATCH_SIZE, BUFFER_TIMEOUT);
```
Comment (dinujoh) [inline data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java:236]: Why is this added instead of creating on constructor ?

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java [hunk 3]
```diff
@@ -82,15 +89,16 @@ public BinlogEventListener(final Buffer<Record<Event>> buffer,
...
         this.binaryLogClient = binaryLogClient;
         tableMetadataMap = new HashMap<>();
         recordConverter = new StreamRecordConverter(sourceConfig.getStream().getPartitionCount());
-        bufferAccumulator = BufferAccumulator.create(buffer, DEFAULT_BUFFER_BATCH_SIZE, BUFFER_TIMEOUT);
         s3Prefix = sourceConfig.getS3Prefix();
         tableNames = sourceConfig.getTableNames();
         isAcknowledgmentsEnabled = sourceConfig.isAcknowledgmentsEnabled();
         this.pluginMetrics = pluginMetrics;
         pipelineEvents = new ArrayList<>();
+        executorService = Executors.newCachedThreadPool();
```
Comment (dlvenable) [inline data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java:101]: I think we want to have an upper bound on this. And make it configurable.

Please use `BackgroundThreadFactory` to make tracking threads easier and promote shutdown.

Example:

https://github.com/opensearch-project/data-prepper/blob/afe84648d662ddc469a6cbf9f22ccd6ff9dfda2c/data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/SqsService.java#L52

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java [hunk 4]
```diff
@@ -51,6 +54,7 @@ public class BinlogEventListener implements BinaryLogClient.EventListener {
     static final String CHANGE_EVENTS_PROCESSING_ERROR_COUNT = "changeEventsProcessingErrors";
     static final String BYTES_RECEIVED = "bytesReceived";
     static final String BYTES_PROCESSED = "bytesProcessed";
+    static final String REPLICATION_LOG_EVENT_PROCESSING_TIME = "replicationLogEventProcessingTime";
```
Comment (dlvenable) [inline data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java:57]: Is there another name that would be appropriate instead of "event" here? This can be confusing with Data Prepper Events.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java (hunk 2 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java:236]: As @dlvenable mentioned below, it's not thread safe.

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java (hunk 4 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java:57]: Agreed. I can change it to "record", "item", or "entry" to differentiate. How about "replicationLogEntryProcessingTime"?

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java (hunk 1 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java]: Updated.

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java (hunk 3 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java]: Done.

**Review by dinujoh — APPROVED** [review]

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java (hunk 4 above)
Comment (dlvenable) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java]: I like that term - "entry".
